src/controller/CommissioningWindowParams.h

```diff
@@ -0,0 +1,185 @@
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <lib/core/CHIPCallback.h>
+#include <lib/core/CHIPError.h>
+#include <lib/core/NodeId.h>
+#include <lib/core/Optional.h>
+#include <lib/support/Span.h>
+#include <setup_payload/SetupPayload.h>
+#include <system/SystemClock.h>
+
+namespace chip {
+namespace Controller {
+
+// Passing SetupPayload by value on purpose, in case a consumer decides to reuse
+// this object from inside the callback.
+typedef void (*OnOpenCommissioningWindow)(void * context, NodeId deviceId, CHIP_ERROR status, SetupPayload payload);
+typedef void (*OnOpenCommissioningWindowWithVerifier)(void * context, NodeId deviceId, CHIP_ERROR status);
+typedef void (*OnOpenBasicCommissioningWindow)(void * context, NodeId deviceId, CHIP_ERROR status);
+
+template <typename Derived>
+class CommissioningWindowCommonParams
+{
+public:
+    CommissioningWindowCommonParams() = default;
+
+    NodeId GetNodeId() const { return mNodeId; }
+    // The node identifier of device
+    Derived & SetNodeId(NodeId nodeId)
+    {
+        mNodeId = nodeId;
+        return static_cast<Derived &>(*this);
+    }
+
+    System::Clock::Seconds16 GetTimeout() const { return mTimeout; }
+    // The duration for which the commissioning window should remain open.
+    Derived & SetTimeout(System::Clock::Seconds16 timeout)
+    {
+        mTimeout = timeout;
+        return static_cast<Derived &>(*this);
+    }
+    Derived & SetTimeout(uint16_t timeoutSeconds) { return SetTimeout(System::Clock::Seconds16(timeoutSeconds)); }
+
+    // The PAKE iteration count associated with the PAKE Passcode ID and
+    // ephemeral PAKE passcode verifier to be used for this commissioning.
+    uint32_t GetIteration() const { return mIteration; }
+    Derived & SetIteration(uint32_t iteration)
+    {
+        mIteration = iteration;
+        return static_cast<Derived &>(*this);
+    }
+
+    // The long discriminator for the DNS-SD advertisement.
+    uint16_t GetDiscriminator() const { return mDiscriminator; }
+    Derived & SetDiscriminator(uint16_t discriminator)
+    {
+        mDiscriminator = discriminator;
+        return static_cast<Derived &>(*this);
+    }
+
+private:
+    NodeId mNodeId                    = 0;
+    System::Clock::Seconds16 mTimeout = System::Clock::Seconds16(300);
+    uint32_t mIteration               = 1000;
+    uint16_t mDiscriminator           = 0;
+};
+
+class CommissioningWindowPasscodeParams : public CommissioningWindowCommonParams<CommissioningWindowPasscodeParams>
+{
+public:
+    CommissioningWindowPasscodeParams() = default;
+
+    bool HasSetupPIN() const { return mSetupPIN.HasValue(); }
+    // Get the value of salt if present.
+    // Returns 0 if absent, make sure to check HasSetupPIN() if a valid value is required.
```

Comment:
Why not just go ahead and crash on misuse by calling `Value()`, not `ValueOr()`?

Reply:
💥 https://github.com/project-chip/connectedhomeip/pull/34209/commits/da28dd643e7419092f7a073ab2011fda251f2504